csharp/src/Ice/SocketStream.cs

```diff
@@ -383,6 +401,20 @@ internal virtual async ValueTask SendInitializeFrameAsync(CancellationToken canc
             // The default implementation only supports the Ice2 protocol
             Debug.Assert(_socket.Endpoint.Protocol == Protocol.Ice2);
 
+            if (frame.Size > _socket.OutgoingFrameMaxSize)
+            {
+                if (frame is OutgoingRequestFrame)
+                {
+                    throw new MarshalException($"the request size is larger than {_socket.OutgoingFrameMaxSize}");
```

Comment:
I think this should be a local exception. Then, when forwarding requests (e.g. ForwardAsync), the forwarder can convert some local exceptions into remote exceptions. See comment on MarshalException.